File: 17. Letter Combinations of a Phone Number/Recursive Writing/letter_combinations_of_a_phone_number.php
<?php
$combinations= [" ", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"];

function wordFromOdo($relevantDigits, $digitsOdo){
	$out = "";
	global $combinations;
	for ($i = 0; $i < strlen($relevantDigits); $i++){
		$out .= $combinations[(int)$relevantDigits[$i]][$digitsOdo[$i]];
	}
	
	return ($out);
}

function wordFill($relevantDigits, $digitsOdo, $depth, $maxDepth, &$sampleSpace){
	global $combinations;
	$letterOptions = $combinations[(int)$relevantDigits[$depth]];
	$lenLetterOptions = strlen($letterOptions);
	
	for ($i = 0; $i < $lenLetterOptions; $i++){
		$digitsOdo[$depth] = $i;
		if ($depth < ($maxDepth - 1)){
			wordFill($relevantDigits, $digitsOdo, ($depth + 1), $maxDepth, $sampleSpace);
		} else {
			//echo (wordFromOdo($relevantDigits, $digitsOdo) . "\n");
			array_push($sampleSpace, wordFromOdo($relevantDigits, $digitsOdo));
		}
	}
}

function getLetterCombinations ($digits){
	global $combinations;
	$out = [];
	
	$lenDigits = strlen($digits);
	$odo = [];
	
	$useableDigits = "";
	
	for ($i = 0; $i < $lenDigits; $i++){
		//echo($combinations[(int)$digits[$i]]. "\n");
		if (strlen($combinations[(int)$digits[$i]]) == 0){
			continue;
		}
		
		$useableDigits .= $digits[$i];
		array_push($odo, 0);
	}
	
	$lenUseableDigits = strlen($useableDigits);
	
	wordFill($useableDigits, $odo, 0, $lenUseableDigits, $out);
	
	return ($out);
}

$digits = "6872";
echo (json_encode(getLetterCombinations($digits)) . "\n\n\n\n");
?>

File: 17. Letter Combinations of a Phone Number/Recursive Writing/letter_combinations_of_a_phone_number.rb
$combinations= [" ", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]

def word_from_odo(relevant_digits, digits_odo)
	out = ""
	
	for i in 0...relevant_digits.length
		out += $combinations[Integer(relevant_digits[i])][digits_odo[i]];
	end
	
	return out
end

def word_fill (relevant_digits, digits_odo, depth, max_depth, sample_space)
	letter_options = $combinations[Integer(relevant_digits[depth])]
	len_letter_options = letter_options.length;
	
	for i in 0...len_letter_options
		digits_odo[depth] = i
		if (depth < (max_depth - 1))
			word_fill(relevant_digits, digits_odo, (depth + 1), max_depth, sample_space)
		else
			sample_space << (word_from_odo(relevant_digits, digits_odo))
		end
	end
end

def get_letter_combinations(digits)
	out = []
	
	len_digits = digits.length
	odo = []
	
	useable_digits = ""
	
	for i in 0...len_digits
		if ($combinations[Integer(digits[i])].length == 0)
			next
		end
		
		useable_digits += digits[i]
		odo << 0
	end
	
	len_useable_digits = useable_digits.length
	word_fill(useable_digits, odo, 0, len_useable_digits, out)
	
	return out
end

digits = "6872"
puts(get_letter_combinations(digits))

